### PR TITLE
[export] Dynamo symint support

### PR DESCRIPTION
**Pull request by angelayi**
Basically adds native _IntWrapper support to dynamo. Here's my process of trying to make symint input support work on dynamo, and how I ended up with this approach [(doc)](https://docs.google.com/document/d/1GvNRQd8BnxlMay_hrEVgEta6VUeUW_hcFeRuB7q1nDY/edit?tab=t.0).

What I did was, before passing inputs to dynamo.export, I first wrap them with a class, `_IntWrapper`. When processing dynamic shapes, I will then add the corresponding dynamic shape specification to the `dynamism` field stored on the `_IntWrapper`. If there is no dynamism specified, then this will get unwrapped back to an integer. When dynamo tracing, when we encounter an `_IntWrapper`, we will convert this to a symint if the dynamism was specified as `Dim.DYNAMIC/AUTO`. Dynamo will then trace a graph that contains symint inputs, which will get passed to AOTAutograd and so on.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -1444,6 +1447,23 @@ def build_key_value(i, k, v):
...
+        elif IntWrapperVariable.is_matching_object(value):
+            from torch.export.dynamic_shapes import _DimHintType
+
+            if value.dynamism is None or value.dynamism.type == _DimHintType.STATIC:
+                return self.wrap_symint(value.val)
+            elif value.dynamism.type == _DimHintType.DYNAMIC:
+                return self.wrap_symint(
+                    value.val,
+                    dynamism=DimDynamic.DYNAMIC,
+                    context=SymIntSymbolicContext(
```
Comment (angelayi) [inline torch/_dynamo/variables/builder.py:1459]: this context allows us to throw ConstraintViolationErrors if the symint is specialized

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -2893,23 +2918,6 @@ def is_dynamic_source(source_name: str) -> bool:
     return False
 
 
-# Tracks the sources of all fake tensors we wrap in Dynamo.
-# Used by shape guard computation.
-@dataclasses.dataclass
-class TrackedFake:
```
Comment (angelayi) [inline torch/_dynamo/variables/builder.py:2899]: moved this to symbolic_shapes.py to prevent circular imports since export uses this directly

File: torch/_export/utils.py [hunk 3]
```diff
@@ -308,7 +308,13 @@ def _check_symint(
     keypath: KeyPath,
     i: Optional[int] = None,
 ) -> None:
-    if isinstance(arg, torch.SymInt) and not arg.node.expr.is_number:
+    from torch.export.dynamic_shapes import _IntWrapper
+
+    if (
+        isinstance(arg, torch.SymInt)
+        and not arg.node.expr.is_number
+        or isinstance(arg, _IntWrapper)
```
Comment (angelayi) [inline torch/_export/utils.py:316]: we run into this when retracing, so just skip over these checks

File: torch/export/_trace.py [hunk 4]
```diff
@@ -736,6 +764,11 @@ def _export_to_torch_ir(
         )
 
     kwargs = kwargs or {}
+
+    # Map ints to a wrapper structure to help us mark it as dynamic, if it is
+    # dynamic. We will unwrap ints in fakify later.
+    args, kwargs = pytree.tree_map_only(int, lambda a: _IntWrapper(a), (args, kwargs))
```
Comment (Skylion007) [inline torch/export/_trace.py:770]:
```suggestion
    args, kwargs = pytree.tree_map_only(int, _IntWrapper, (args, kwargs))

File: torch/export/_trace.py [hunk 5]
```diff
@@ -744,6 +794,7 @@ def _export_to_torch_ir(
                 gm_torch_level, _ = torch._dynamo.export(
                     f,
                     dynamic_shapes=dynamic_shapes,  # type: ignore[arg-type]
+                    constraints=constraints,  # type: ignore[arg-type]
```
Comment (pianpwk) [inline torch/export/_trace.py:797]: just wondering, we could have done this in _dynamo.export too, and avoided passing constraints in? not that it's a big deal

File: torch/export/_trace.py [hunk 6]
```diff
@@ -730,8 +760,28 @@ def _export_to_torch_ir(
...
+    # containing just integer computation, dynamo will run its generated
+    # bytecode with these args/kwargs, which will error because we cannot
+    # directly apply int operations on IntWrapper. So we will just unwrap
+    # them here.
+    args, kwargs = pytree.tree_map_only(
+        _IntWrapper,
+        lambda a: a.val
+        if a.dynamism is None or a.dynamism.type == _DimHintType.STATIC
+        else a,
+        (args, kwargs),
```
Comment (pianpwk) [inline torch/export/_trace.py:782]: nit: could we kept process_dynamic_shapes in _dynamo.export, and not passed constraints here?

also could we have not wrapped these static ints to start (with export dynamic_shapes's version of tree_map)

not that it's a big deal

File: torch/export/_trace.py [hunk 7]
```diff
@@ -220,40 +226,64 @@ def _rewrite_tracepoint_node(gm: torch.fx.GraphModule):
...
+            fake_vals.append(node.meta.get("example_value"))
+
+    detected_fake_mode = detect_fake_mode(fake_inps + fake_vals)
+    detected_shape_env = detect_shape_env(fake_inps + fake_vals)
+
+    if detected_fake_mode:
+        if detected_shape_env:
+            assert (
+                detected_shape_env is detected_fake_mode.shape_env
+            ), "Detected shape env does not match fake mode's shape env"
```
Comment (pianpwk) [inline torch/export/_trace.py:270]: hmm, why is this a problem now?

Comment (angelayi) [reply torch/export/_trace.py:270]: Oh it's just some consistency checking between the shape env in the symint inputs and the shape env attached to the fake mode. Because now we can have inputs that are both symints and fake tensors.

File: torch/export/_trace.py (hunk 6 above)
Comment (angelayi) [reply torch/export/_trace.py:782]: > could we kept process_dynamic_shapes in _dynamo.export, and not passed constraints here?

The reason why I needed to pass in constraints here is because there's a [tested behavior](https://github.com/pytorch/pytorch/blob/3cd69350edded21849f0b3724edc1fb87df563b3/test/dynamo/test_export.py#L287) in dynamo.export which checks that any side effects that occur while tracing *should* be propagated to the input. The specific test checks that if the model modifies a value in an input list then after tracing the input list should be modified.

However if we do this tree_map_only step, even if none of the values are changed, it still creates a new list. So then when we dynamo trace, side effects will only change the pytree created list but not the original input.

I feel like this (the side effect propagation) is not a behavior we support in torch.export (esp since non-strict can't support this) so I just moved this pytree code to the torch.export level and kept the behavior in the dynamo.export level. It makes sense that dynamo has this behavior though, since in the torch.compile case we do want to propagate the side effects.

> also could we have not wrapped these static ints to start 

I guess we could but process_dynamic_shapes is just to process the dynamic shapes.. feels a little weird that it would also return a new set of arguments...

File: torch/export/_trace.py (hunk 7 above)
Comment (pianpwk) [reply torch/export/_trace.py:270]: could we just add the fake mode detection for SymInts to detect_fake_mode, or does that introduce new failures? This is a bit confusing.

Comment (pianpwk) [reply torch/export/_trace.py:270]: It seems the main reason we can't add this logic to detect_fake_mode is because we can't access a fake mode from the ShapeEnv, so we can't compare ownership between tensors & SymInts? Could we add a comment here? It looks pretty confusing to have both calls.

File: torch/export/_trace.py (hunk 6 above)
Comment (pianpwk) [reply torch/export/_trace.py:782]: > I guess we could but process_dynamic_shapes is just to process the dynamic shapes.. feels a little weird that it would also return a new set of arguments...

nit: oh I meant change line 766 to wrap only ints specified with AUTO or DYNAMIC

File: torch/export/_trace.py (hunk 7 above)
Comment (angelayi) [reply torch/export/_trace.py:270]: > is because we can't access a fake mode from the ShapeEnv, so we can't compare ownership between tensors & SymInts? 

Ya sounds good I'll add a comment

File: torch/export/_trace.py (hunk 6 above)
Comment (angelayi) [reply torch/export/_trace.py:782]: > oh I meant change line 766 to wrap only ints specified with AUTO or DYNAMIC

mm I can't think of an easy way of doing this since to map inputs to the dynamic_shapes you have to do the combine args, and then tree_map_with_path... maybe a skill issue

**Review by pianpwk — APPROVED** [review]
please add the comments
